### ISSUE TITLE
Use `goose` for database migrations

### ISSUE THREAD

**Issue by erikgrinaker** [sanity-io/gradient#519]
I suggest we use [Goose](https://github.com/pressly/goose) for database migrations, rather than rolling our own. Uses simple SQL text files for migrations, and has a feature-complete command-line tool as well as an API.

**Comment by atombender** [comment]
I disagree here. First of all, the data store abstracts the underlying implementation (Postgres), and so the consumer of Gradient shouldn't need to know what an "SQL migration" is. The data store itself exposes an API to upgrade itself and executes the necessary migrations, and will abort if migrations are required. This is similar to other datastores such as Elasticsearch where the internal storage mechanism is a private implementation detail.

In fact, I'd like to go one step further and suggest that we explicitly version the data store. Currently, our migrations are just named, ordered scripts. We should introduce a "data store format version" that we track, so that we can err with something like:

> Data store version 3 found; this version of Gradient requires an upgrade to version 4. Run "gradientadm upgrade" to perform upgrade.

Secondly, our migrations are pretty simple. We have a pretty limited surface area we need to cover on the data store side.

I'd love to use a third-party library to manage migrations as files, but Goose is a command-line tool and not a library, so I don't think want to use it.

**Comment by erikgrinaker** [comment]
> The data store itself exposes an API to upgrade itself and executes the necessary migrations, and will abort if migrations are required.

Shouldn't it ideally handle ES index migrations as well then?

>I'd love to use a third-party library to manage migrations as files, but Goose is a command-line tool and not a library, so I don't think want to use it.

Goose is a library too, the command-line tool is a consumer.

**Comment by atombender** [comment]
> Shouldn't it ideally handle ES index migrations as well then?

Those aren't related to the data store. But yes, we should do it the same way. We have automatic migrations today, I believe. If you increment the version number field, it should cause new indexes to be built while the old ones stick around.

We could look into using Goose as a library.
